From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Manish Bhardwaj <mbhardwaj@nvidia.com>
Date: Sun, 6 Jul 2025 09:45:13 +0000
Subject: [PATCH 192/241] ivc: fix comments raised by secure code inspection

Bug 5099275

Change-Id: Id37aa1f45bdcec04e560fafd06e3899439eca361
Signed-off-by: Manish Bhardwaj <mbhardwaj@nvidia.com>
(cherry picked from commit d34b3a7560365685b997af51c96c5a71abac4d7c)
Reviewed-on: https://git-master.nvidia.com/r/c/3rdparty/canonical/linux-noble/+/3449489
GVS: buildbot_gerritrpt <buildbot_gerritrpt@nvidia.com>
Reviewed-by: svcacv <svcacv@nvidia.com>
Reviewed-by: Sandeep Trasi <strasi@nvidia.com>
Reviewed-by: Suresh Venkatachalam <skathirampat@nvidia.com>
(cherry picked from commit 2aea973e8b1aec28a4609e5119824cb3d4b00846)

diff --git a/drivers/firmware/tegra/ivc.c b/drivers/firmware/tegra/ivc.c
index d3bc1d435b2f5..e4745c08ff8fa 100644
--- a/drivers/firmware/tegra/ivc.c
+++ b/drivers/firmware/tegra/ivc.c
@@ -649,6 +649,12 @@ int tegra_ivc_read(struct tegra_ivc *ivc, void __user *usr_buf, void *buf, size_
 	if (buf) {
 		iosys_map_memcpy_from(buf, &map, 0, max_read);
 	} else if (usr_buf) {
+		/* Validate user buffer */
+		if (!access_ok(usr_buf, max_read)) {
+			pr_err("IVC read failed: invalid user buffer\n");
+			return -EFAULT;
+		}
+
 		// FIXME handle io address space
 		if (WARN_ON(map.is_iomem) || copy_to_user(usr_buf, map.vaddr, max_read))
 			return -EFAULT;
@@ -676,6 +682,11 @@ int tegra_ivc_read_peek(struct tegra_ivc *ivc, void __user *usr_buf,
 	if (err)
 		return err;
 
+	/* Ensure offset + size does not exceed buffer bounds */
+	if (offset + size > ivc->frame_size) {
+		return -EINVAL;
+	}
+
 	/* update the buffer with read data*/
 	if (buf) {
 		iosys_map_memcpy_from(buf, &map, offset, size);
@@ -702,6 +713,11 @@ int tegra_ivc_write(struct tegra_ivc *ivc, const void __user *usr_buf, const voi
 	if (err)
 		return err;
 
+	/* Ensure size does not exceed buffer bounds */
+	if (size > ivc->frame_size) {
+		return -EINVAL;
+	}
+
 	/* update the write frame with data buffer*/
 	if (buf) {
 		iosys_map_memcpy_to(&map, 0, buf, size);
-- 
2.43.0

